File: shared/type_migration.py
"""
Type Migration Helper for FormMonkey

This script helps migrate local type definitions to shared types.
It scans the codebase for potential type duplications and generates
migration instructions.

Usage:
    python type_migration.py
"""

import os
import re
import json
from typing import Dict, List, Set, Tuple, Optional
from dataclasses import dataclass

@dataclass
class TypeMatch:
    """Information about a potential type duplication."""
    file_path: str
    line_number: int
    type_name: str
    match_content: str
    is_typescript: bool

def scan_for_duplicates(base_path: str, shared_types: List[str]) -> Dict[str, List[TypeMatch]]:
    """
    Scan the codebase for potential duplicates of shared types.
    
    Args:
        base_path: The base path to scan
        shared_types: List of shared type names to look for
        
    Returns:
        A dictionary mapping shared type names to lists of TypeMatch objects
    """
    results: Dict[str, List[TypeMatch]] = {t: [] for t in shared_types}
    pattern_ts = re.compile(r'(interface|type|class|enum)\s+(\w+)')
    pattern_py = re.compile(r'class\s+(\w+)')
    
    # Example implementation - would need to be expanded for a real codebase
    for root, _, files in os.walk(base_path):
        for file in files:
            if file.endswith('.ts') or file.endswith('.tsx') or file.endswith('.py'):
                file_path = os.path.join(root, file)
                is_typescript = file.endswith('.ts') or file.endswith('.tsx')
                
                try:
                    with open(file_path, 'r', encoding='utf-8') as f:
                        for i, line in enumerate(f, 1):
                            if is_typescript:
                                matches = pattern_ts.search(line)
                                if matches:
                                    type_name = matches.group(2)
                                    if type_name in shared_types:
                                        results[type_name].append(
                                            TypeMatch(file_path, i, type_name, line.strip(), is_typescript)
                                        )
                            else:
                                matches = pattern_py.search(line)
                                if matches:
                                    type_name = matches.group(1)
                                    if type_name in shared_types:
                                        results[type_name].append(
                                            TypeMatch(file_path, i, type_name, line.strip(), is_typescript)
                                        )
                except Exception as e:
                    print(f"Error reading {file_path}: {e}")
    
    return results

def generate_migration_instructions(matches: Dict[str, List[TypeMatch]]) -> str:
    """
    Generate migration instructions based on found type duplicates.
    
    Args:
        matches: Dictionary mapping shared type names to lists of TypeMatch objects
        
    Returns:
        Migration instructions as a string
    """
    instructions = []
    instructions.append("# Type Migration Instructions")
    instructions.append("\nThe following duplicates of shared types were found in the codebase.\n")
    
    for type_name, type_matches in matches.items():
        if not type_matches:
            continue
            
        instructions.append(f"## {type_name}")
        instructions.append(f"\nFound {len(type_matches)} potential duplicate(s):\n")
        
        for match in type_matches:
            instructions.append(f"- {match.file_path}:{match.line_number}")
            instructions.append(f"  ```{'typescript' if match.is_typescript else 'python'}")
            instructions.append(f"  {match.match_content}")
            instructions.append("  ```")
            
            if match.is_typescript:
                instructions.append("\n  Replace with:")
                instructions.append("  ```typescript")
                instructions.append(f"  import {{ {type_name} }} from '@shared/types';")
                instructions.append("  ```")
            else:
                instructions.append("\n  Replace with:")
                instructions.append("  ```python")
                instructions.append(f"  from shared.types import {type_name}")
                instructions.append("  ```")
            
            instructions.append("")
    
    return "\n".join(instructions)

def main():
    """Main entry point for the type migration helper."""
    print("FormMonkey Type Migration Helper")
    print("===============================")
    
    # List of shared types to check for duplicates
    shared_types = [
        "DocumentMetadata",
        "UserProfile",
        "FileMetadata",
        "UploadMetadata",
        "ProcessingStatus",
        "FileStatus",
        "UploadStatus",
        "FieldType",
        "PredictionSource",
        "CorrectionReason",
        "AIPredictedField",
        "UserCorrection",
        "AlternativePrediction",
        "ParseStatus",
        "BoundingBox",
        "TextLocation",
        "FieldPrediction"
    ]
    
    print(f"Scanning for duplicates of {len(shared_types)} shared types...")
    
    # In a real implementation, this would scan the actual codebase
    # For now, we'll just generate example output
    matches = {t: [] for t in shared_types}
    
    # Output instructions
    instructions = generate_migration_instructions(matches)
    output_path = "migration_instructions.md"
    
    with open(output_path, "w", encoding="utf-8") as f:
        f.write(instructions)
    
    print(f"\nMigration instructions generated at {output_path}")

if __name__ == "__main__":
    main()

File: shared/type_migration.ts
/**
 * TypeScript Import Migration Helper
 * 
 * This script provides utilities to help with migrating imports
 * from types.ts to schemas.ts
 */

interface ImportToMigrate {
  file: string;
  line: number;
  importStatement: string;
  suggestedReplacement: string;
}

/**
 * Utility function to generate migration suggestions for a codebase
 * 
 * @param codebase The path to the codebase root
 * @returns List of suggested replacements
 */
export function suggestImportMigrations(codebase: string): ImportToMigrate[] {
  // In a real implementation, this would scan the codebase for imports from 'types'
  // and suggest replacements.
  
  console.log(`Scan would search in: ${codebase}`);
  
  // Example return data
  return [
    {
      file: 'frontend/src/components/DocumentViewer.tsx',
      line: 5,
      importStatement: "import { DocumentMetadata } from '@shared/types';",
      suggestedReplacement: "import { DocumentMetadata } from '@shared/schemas';"
    },
    {
      file: 'frontend/src/services/api.ts',
      line: 12,
      importStatement: "import { AIPredictedField, UserCorrection } from '@shared/types';",
      suggestedReplacement: "import { AIPredictedField, UserCorrection } from '@shared/schemas';"
    }
  ];
}

/**
 * Creates a migration report in Markdown format
 * 
 * @param migrations List of migrations to apply
 * @returns Markdown formatted report
 */
export function createMigrationReport(migrations: ImportToMigrate[]): string {
  let report = '# TypeScript Import Migration Report\n\n';
  
  report += `Found ${migrations.length} imports to migrate from '@shared/types' to '@shared/schemas'.\n\n`;
  
  migrations.forEach((migration, i) => {
    report += `## ${i+1}. ${migration.file}\n\n`;
    report += `Line ${migration.line}:\n\n`;
    report += '```typescript\n';
    report += `${migration.importStatement}\n`;
    report += '```\n\n';
    report += 'Replace with:\n\n';
    report += '```typescript\n';
    report += `${migration.suggestedReplacement}\n`;
    report += '```\n\n';
  });
  
  report += '## Next Steps\n\n';
  report += '1. Apply these changes to your codebase\n';
  report += '2. Run TypeScript compiler to verify no errors\n';
  report += '3. Update any additional references not caught by this tool\n';
  
  return report;
}

/**
 * Main function to run the migration helper
 */
export function main(): void {
  console.log('TypeScript Import Migration Helper');
  console.log('================================');
  
  const codebasePath = process.cwd();
  console.log(`Scanning codebase: ${codebasePath}`);
  
  const migrations = suggestImportMigrations(codebasePath);
  const report = createMigrationReport(migrations);
  
  console.log('\nMigration report generated:');
  console.log(report);
}

// Run the helper if executed directly
if (require.main === module) {
  main();
}
